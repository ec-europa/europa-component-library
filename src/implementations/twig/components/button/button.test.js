import {
  merge,
  renderTwigFileAsNode,
  renderTwigFileAsHtml,
} from '@ecl/test-utils';
import { axe, toHaveNoViolations } from 'jest-axe';

// Import data for tests
import dataPrimary from '@ecl/specs-component-button/demo/data--primary';
import dataSecondary from '@ecl/specs-component-button/demo/data--secondary';
import dataCall from '@ecl/specs-component-button/demo/data--call';
import dataGhost from '@ecl/specs-component-button/demo/data--ghost';

expect.extend(toHaveNoViolations);

describe('Button', () => {
  const template = '@ecl/button/button.html.twig';
  const render = (params) => renderTwigFileAsNode(template, params);

  describe('Primary', () => {
    test('renders correctly', () => {
      expect.assertions(1);

      return expect(render(dataPrimary)).resolves.toMatchSnapshot();
    });

    test(`passes the accessibility tests`, async () => {
      expect(
        await axe(renderTwigFileAsHtml(template, dataPrimary))
      ).toHaveNoViolations();
    });
  });

  describe('Secondary', () => {
    test('renders correctly', () => {
      expect.assertions(1);

      return expect(render(dataSecondary)).resolves.toMatchSnapshot();
    });
  });

  describe('CTA', () => {
    test('renders correctly', () => {
      expect.assertions(1);

      return expect(render(dataCall)).resolves.toMatchSnapshot();
    });

    test(`passes the accessibility tests`, async () => {
      expect(
        await axe(renderTwigFileAsHtml(template, dataCall))
      ).toHaveNoViolations();
    });
  });

  describe('Text', () => {
    test('renders correctly', () => {
      expect.assertions(1);

      return expect(render(dataGhost)).resolves.toMatchSnapshot();
    });

    test(`passes the accessibility tests`, async () => {
      expect(
        await axe(renderTwigFileAsHtml(template, dataGhost))
      ).toHaveNoViolations();
    });
  });

  describe('CTA button - icon before', () => {
    const buttonData = {
      label: 'CTA Button with icon',
      variant: 'cta',
      icon_position: 'before',
      icon: {
        path: 'static/icons.svg',
        name: 'corner-arrow',
        size: 'fluid',
      },
    };

    test('renders correctly', () => {
      expect.assertions(1);

      return expect(render(buttonData)).resolves.toMatchSnapshot();
    });

    test(`passes the accessibility tests`, async () => {
      expect(
        await axe(renderTwigFileAsHtml(template, dataCall))
      ).toHaveNoViolations();
    });
  });

  describe('CTA button - icon after', () => {
    const buttonData = {
      label: 'CTA Button with icon',
      variant: 'cta',
      icon: {
        path: 'static/icons.svg',
        name: 'corner-arrow',
        size: 'fluid',
      },
    };

    test('renders correctly', () => {
      expect.assertions(1);

      return expect(render(buttonData)).resolves.toMatchSnapshot();
    });

    test('renders correctly with extra class names', () => {
      expect.assertions(1);

      const withExtraClasses = merge(buttonData, {
        extra_classes: 'custom-class custom-class--test',
      });

      return expect(render(withExtraClasses)).resolves.toMatchSnapshot();
    });

    test('renders correctly with extra attributes', () => {
      expect.assertions(1);

      const withExtraAttributes = merge(buttonData, {
        extra_attributes: [
          { name: 'data-test', value: 'data-test-value' },
          { name: 'data-test-1', value: 'data-test-value-1' },
        ],
      });

      return expect(render(withExtraAttributes)).resolves.toMatchSnapshot();
    });
  });
});
