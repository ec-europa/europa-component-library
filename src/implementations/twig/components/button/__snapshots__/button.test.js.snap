// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button CTA button - icon after renders correctly 1`] = `
<jest>
  <button
    class="ecl-button ecl-button--cta"
    type="submit"
  >
    <span
      class="ecl-button__container"
    >
      <span
        class="ecl-button__label"
        data-ecl-label="true"
      >
        CTA Button with icon
      </span>
      <svg
        aria-hidden="true"
        class="ecl-icon ecl-icon--fluid ecl-button__icon ecl-button__icon--after"
        data-ecl-icon=""
        focusable="false"
      >
        <use
          xlink:href="static/icons.svg#corner-arrow"
        />
      </svg>
    </span>
  </button>
</jest>
`;

exports[`Button CTA button - icon after renders correctly with extra attributes 1`] = `
<jest>
  <button
    class="ecl-button ecl-button--cta"
    data-test="data-test-value"
    data-test-1="data-test-value-1"
    type="submit"
  >
    <span
      class="ecl-button__container"
    >
      <span
        class="ecl-button__label"
        data-ecl-label="true"
      >
        CTA Button with icon
      </span>
      <svg
        aria-hidden="true"
        class="ecl-icon ecl-icon--fluid ecl-button__icon ecl-button__icon--after"
        data-ecl-icon=""
        focusable="false"
      >
        <use
          xlink:href="static/icons.svg#corner-arrow"
        />
      </svg>
    </span>
  </button>
</jest>
`;

exports[`Button CTA button - icon after renders correctly with extra class names 1`] = `
<jest>
  <button
    class="ecl-button ecl-button--cta custom-class custom-class--test"
    type="submit"
  >
    <span
      class="ecl-button__container"
    >
      <span
        class="ecl-button__label"
        data-ecl-label="true"
      >
        CTA Button with icon
      </span>
      <svg
        aria-hidden="true"
        class="ecl-icon ecl-icon--fluid ecl-button__icon ecl-button__icon--after"
        data-ecl-icon=""
        focusable="false"
      >
        <use
          xlink:href="static/icons.svg#corner-arrow"
        />
      </svg>
    </span>
  </button>
</jest>
`;

exports[`Button CTA button - icon before renders correctly 1`] = `
<jest>
  <button
    class="ecl-button ecl-button--cta"
    type="submit"
  >
    <span
      class="ecl-button__container"
    >
      <svg
        aria-hidden="true"
        class="ecl-icon ecl-icon--fluid ecl-icon--rotate-90 ecl-button__icon ecl-button__icon--before"
        data-ecl-icon=""
        focusable="false"
      >
        <use
          xlink:href="static/icons.svg#corner-arrow"
        />
      </svg>
      <span
        class="ecl-button__label"
        data-ecl-label="true"
      >
        Call to action button
      </span>
    </span>
  </button>
</jest>
`;

exports[`Button CTA renders correctly 1`] = `
<jest>
  <button
    class="ecl-button ecl-button--cta"
    type="submit"
  >
    <span
      class="ecl-button__container"
    >
      <span
        class="ecl-button__label"
        data-ecl-label="true"
      >
        Call to action button
      </span>
      <svg
        aria-hidden="true"
        class="ecl-icon ecl-icon--fluid ecl-icon--rotate-90 ecl-button__icon ecl-button__icon--after"
        data-ecl-icon=""
        focusable="false"
      >
        <use
          xlink:href="static/icons.svg#corner-arrow"
        />
      </svg>
    </span>
  </button>
</jest>
`;

exports[`Button Primary renders correctly 1`] = `
<jest>
  <button
    class="ecl-button ecl-button--primary"
    type="submit"
  >
    Primary button
  </button>
</jest>
`;

exports[`Button Secondary renders correctly 1`] = `
<jest>
  <button
    class="ecl-button ecl-button--secondary"
    type="button"
  >
    Secondary button
  </button>
</jest>
`;

exports[`Button Text renders correctly 1`] = `
<jest>
  <button
    class="ecl-button ecl-button--ghost"
    type="button"
  >
    Ghost button
  </button>
</jest>
`;
